syntax = "proto3";

package store.v1;

import "rbt/v1alpha1/options.proto";

////////////////////////////////////////////////////////////////////////

message Cart {
  option (rbt.v1alpha1.state) = {};

  repeated CartItem items = 1;
}

////////////////////////////////////////////////////////////////////////

service CartMethods {
  rpc Create(CreateCartRequest) returns (CreateCartResponse) {
    option (rbt.v1alpha1.method) = {
      writer: {
        constructor: {},
      },
    };
  }
  rpc AddItem(AddItemRequest) returns (AddItemResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
  rpc GetItems(GetItemsRequest) returns (GetItemsResponse) {
    option (rbt.v1alpha1.method).reader = {};
  }
  rpc UpdateItemQuantity(UpdateItemQuantityRequest) returns (UpdateItemQuantityResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
  rpc RemoveItem(RemoveItemRequest) returns (RemoveItemResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
  rpc EmptyCart(EmptyCartRequest) returns (EmptyCartResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
}

////////////////////////////////////////////////////////////////////////

message CartItem {
  string product_id = 1;
  int32 quantity = 2;
  int64 added_at = 3;
  string name = 4;
  uint32 price_cents = 5;
  string picture = 6;
}

message CreateCartRequest {}

message CreateCartResponse {}

message AddItemRequest {
  CartItem item = 2;
  string catalog_id = 3;
}

message AddItemResponse {}

message GetItemsRequest {}

message GetItemsResponse {
  repeated CartItem items = 1;
}

message UpdateItemQuantityRequest {
  string product_id = 2;
  int32 quantity = 3;
}

message UpdateItemQuantityResponse {}

message RemoveItemRequest {
  string product_id = 2;
}

message RemoveItemResponse {}

message EmptyCartRequest {
}

message EmptyCartResponse {}

////////////////////////////////////////////////////////////////////////

message ProductCatalog {
  option (rbt.v1alpha1.state) = {};

  repeated Product products = 1;
}

////////////////////////////////////////////////////////////////////////

service ProductCatalogMethods {
  rpc ListProducts(ListProductsRequest) returns (ListProductsResponse) {
    option (rbt.v1alpha1.method).reader = {};
  }
  rpc GetProduct(GetProductRequest) returns (Product) {
    option (rbt.v1alpha1.method).reader = {};
  }
  rpc SearchProducts(SearchProductsRequest) returns (SearchProductsResponse) {
    option (rbt.v1alpha1.method).reader = {};
  }
  rpc AddProduct(AddProductRequest) returns (AddProductResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
  rpc UpdateProduct(UpdateProductRequest) returns (UpdateProductResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
}

////////////////////////////////////////////////////////////////////////

message Product {
  string id = 1;
  string name = 2;
  string description = 3;
  string picture = 4;
  uint32 price_cents = 5;
  repeated string categories = 6;
  int32 stock_quantity = 7;
}

message ListProductsRequest {
  string catalog_id = 1;
}

message ListProductsResponse {
  repeated Product products = 1;
}

message GetProductRequest {
  string catalog_id = 1;
  string product_id = 2;
}

message SearchProductsRequest {
  string catalog_id = 1;
  string query = 2;
}

message SearchProductsResponse {
  repeated Product products = 1;
}

message AddProductRequest {
  string catalog_id = 1;
  Product product = 2;
}

message AddProductResponse {}

message UpdateProductRequest {
  string catalog_id = 1;
  Product product = 2;
}

message UpdateProductResponse {}

////////////////////////////////////////////////////////////////////////

message Orders {
  option (rbt.v1alpha1.state) = {};

  repeated Order orders = 1;
}

////////////////////////////////////////////////////////////////////////

service OrdersMethods {
  rpc Create(CreateOrdersRequest) returns (CreateOrdersResponse) {
    option (rbt.v1alpha1.method) = {
      writer: {
        constructor: {},
      },
    };
  }
  rpc AddOrder(AddOrderRequest) returns (AddOrderResponse) {
    option (rbt.v1alpha1.method).writer = {};
  }
  rpc GetOrders(GetOrdersRequest) returns (GetOrdersResponse) {
    option (rbt.v1alpha1.method).reader = {};
  }
}

////////////////////////////////////////////////////////////////////////

message Order {
  string order_id = 1;
  repeated CartItem items = 2;
  string transaction_id = 3;
  uint32 subtotal_cents = 4;
  uint32 shipping_cost_cents = 5;
  uint32 total_cents = 6;
  string tracking_number = 7;
  string carrier = 8;
  int64 created_at = 9;
  Address shipping_address = 10;
}

message Address {
  string street_address = 1;
  string city = 2;
  string state = 3;
  string country = 4;
  string zip_code = 5;
}

message CreateOrdersRequest {}

message CreateOrdersResponse {}

message AddOrderRequest {
  Order order = 1;
}

message AddOrderResponse {}

message GetOrdersRequest {}

message GetOrdersResponse {
  repeated Order orders = 1;
}

